pkg/webhook/conversion/conversion.go

```diff
@@ -0,0 +1,250 @@
+	if err != nil {
+		return nil, err
+	}
+	return convertReview, nil
+}
+
+// handles a version conversion request.
+func (wh *Webhook) handleConvertRequest(req *apix.ConversionRequest) (*apix.ConversionResponse, error) {
+	if req == nil {
+		return nil, fmt.Errorf("conversion request is nil")
+	}
+	var objects []runtime.RawExtension
+
+	for _, obj := range req.Objects {
+		src, gvk, err := wh.decoder.Decode(obj.Raw)
+		if err != nil {
+			return nil, err
+		}
+		dst, err := wh.allocateDstObject(req.DesiredAPIVersion, gvk.Kind)
+		if err != nil {
+			return nil, err
+		}
+		err = wh.convertObject(src, dst)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, runtime.RawExtension{Object: dst})
+	}
+	return &apix.ConversionResponse{
+		UID:              req.UID,
+		ConvertedObjects: objects,
+	}, nil
+}
+
+// convertObject will convert given a src object to dst object.
+func (wh *Webhook) convertObject(src, dst runtime.Object) error {
+	srcGVK := src.GetObjectKind().GroupVersionKind()
+	dstGVK := dst.GetObjectKind().GroupVersionKind()
+
+	if srcGVK.GroupKind().String() != dstGVK.GroupKind().String() {
+		return fmt.Errorf("src %T and dst %T does not belong to same API Group", src, dst)
+	}
+
+	if srcGVK.String() == dstGVK.String() {
+		return fmt.Errorf("conversion is not allowed between same type %T", src)
+	}
+
+	srcIsHub, dstIsHub := isHub(src), isHub(dst)
+	srcIsConvertible, dstIsConvertible := isConvertible(src), isConvertible(dst)
+
+	if srcIsHub {
+		if dstIsConvertible {
+			return dst.(conversion.Convertible).ConvertFrom(src.(conversion.Hub))
+		}
+		return fmt.Errorf("%T is not convertible to %T", src, dst)
+	}
+
+	if dstIsHub {
+		if srcIsConvertible {
+			return src.(conversion.Convertible).ConvertTo(dst.(conversion.Hub))
+		}
+		return fmt.Errorf("%T is not convertible %T", dst, src)
+	}
+
+	// neither src nor dst are Hub, means both of them are spoke, so lets get the hub
+	// version type.
+	hub, err := wh.getHub(src)
+	if err != nil {
+		return err
+	}
+
+	// src and dst needs to be convertable for it to work
+	if !srcIsConvertible || !dstIsConvertible {
+		return fmt.Errorf("%T and %T needs to be convertable", src, dst)
+	}
+
+	err = src.(conversion.Convertible).ConvertTo(hub)
+	if err != nil {
+		return fmt.Errorf("%T failed to convert to hub version %T : %v", src, hub, err)
+	}
+
+	err = dst.(conversion.Convertible).ConvertFrom(hub)
+	if err != nil {
+		return fmt.Errorf("%T failed to convert from hub version %T : %v", dst, hub, err)
+	}
+
+	return nil
+}
+
+// getHub returns an instance of the Hub for passed-in object's group/kind.
+func (wh *Webhook) getHub(obj runtime.Object) (conversion.Hub, error) {
+	gvks, _, err := wh.scheme.ObjectKinds(obj)
+	if err != nil {
+		return nil, fmt.Errorf("error retriving object kinds for given object : %v", err)
+	}
+
+	var hub conversion.Hub
+	var isHub, hubFoundAlready bool
+	for _, gvk := range gvks {
+		o, _ := wh.scheme.New(gvk)
```

Comment:
The error should be checked here. If the gvk is not registered in the scheme, `scheme.New` will return an error.

Reply:
Good pt. Added catch.